working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,752 @@
+that it is a _protected_ extension type. A protected extension type can
+declare one or more non-redirecting factory constructors. We use the phrase
+_extension type constructor_ to denote such constructors.
+
+An instance creation expression of the form `Ext<T1, .. Tk>(...)` or
+`Ext<T1, .. Tk>.name(...)` is used to invoke these constructors, and the
+type of such an expression is `Ext<T1, .. Tk>`.
+
+During static analysis of the body of an extension type constructor, the
+return type is considered to be the on-type of the enclosing extension type
+declaration.
+
+It is a compile-time error if it is possible to reach the
+end of an extension type constructor without returning anything. 
+
+*Even in the case where the on-type is nullable and the intended
+representation is the null object, an explicit `return null;` is required.*
+
+A protected extension type is a proper subtype of the top types, and a
+proper supertype of `Never`.
+
+*In particular, there is no subtype relationship between a protected
+extension type and the corresponding on-type.*
+
+When `E` (respectively `E<X1, .. Xk>`) is a protected extension type, and
+`e` is an expression with static type `E` (respectively `E<X1, .. Xk>`), 
+it is a compile-time error to perform a type test (`e is T`) or type cast
+(`e as T`) when the target type `T` is `E` (respectively `E<T1, .. Tk>`).
+
+*The rationale is that an extension type that justifies _any_ constructors
+will need to maintain some invariants, and hence it is not helpful to allow
+implicit introduction of any value of that type with a complete lack of
+enforcement of the invariants.*
+
+*For example:*
+
+```dart
+protected extension type nat on int {
+  factory nat(int value) =>
+      value >= 0 ? value : throw "Attempt to create an invalid nat";
+}
+
+void main() {
+  nat n1 = 42; // Error.
+  var n2 = nat(42); // OK at compile time, and at run time.
+}
+```
+
+*The subtyping relationships are illustrated in the following example:*
+
+```dart
+class IntBox {
+  int i;
+  IntBox(this.i);
+}
+
+protected extension type EvenIntBox on IntBox {
+  factory EvenIntBox(int i) =>
+      i % 2 == 0 ? IntBox(i) : throw "Invalid EvenIntBox";
+  factory EvenIntBox.fromIntBox(IntBox intBox) =>
+      intBox.i % 2 == 0 ? intBox : throw "Invalid EvenIntBox";
+  void next() => this.i += 2;
+}
+
+void main() {
+  var evenIntBox = EvenIntBox(42);
+  evenIntBox.next(); // Methods of `EvenIntBox` maintain the invariant.
+  var intBox = evenIntBox as IntBox; // OK statically and dynamically.
+  intBox.i++; // Invariant of `evenIntBox` violated!
+  evenIntBox = intBox as EvenIntBox; // Error, can't cast to EvenIntBox.
+
+  var evenIntBoxes = [evenIntBox]; // Type `List<EvenIntBox>`.
+  evenIntBoxes[0].next(); // Elements typed as `EvenIntBox`, maintain invariant.
+  List<IntBox> intBoxes = evenIntBoxes; // Compile-time error.
+  intBoxes = evenIntBoxes as dynamic; // Run-time error.
+}
+```
+
+
+### Boxing
```

Comment:
This section seems like an afterthought. I don't buy the motivation, and it distracts from the overall story of zero-overhead abstraction. Consider moving it to the "enhancements' section.

Reply:
Created https://github.com/dart-lang/language/issues/1491 in order to take this discussion separately.